presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt

```diff
@@ -1,61 +1,92 @@
+import com.exallium.mvvmcleanexample.presentation.utils.delegates.DisposableProperty
+import com.exallium.mvvmcleanexample.presentation.utils.distinctObservable
+import com.exallium.mvvmcleanexample.presentation.utils.disposable
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.BaseNotifiableObservable
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.NotifiableObservable
+import com.exallium.mvvmcleanexample.presentation.utils.onNext
 import com.jakewharton.rxrelay2.BehaviorRelay
 import com.jakewharton.rxrelay2.PublishRelay
-import io.reactivex.Observable
-import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
 
-class UserEditViewModel(userEditModel: UserEditModel,
-                        private val router: UiRouter) {
+class UserEditViewModel(
+        private val userEditModel: UserEditModel,
+        private val router: UiRouter,
+        private val firstNameRelay: BehaviorRelay<String> = BehaviorRelay.create(),
+        private val lastNameRelay: BehaviorRelay<String> = BehaviorRelay.create(),
+        private val savedUserRelay: PublishRelay<User> = PublishRelay.create(),
+        disposableProperty: DisposableProperty = disposable(null),
+        notifiableObservable: NotifiableObservable = BaseNotifiableObservable())
+    : NotifiableObservable by notifiableObservable {
 
-    val firstNameError = ObservableField<String>("")
-    val lastNameError = ObservableField<String>("")
+    init {
+        setDelegator(this)
+    }
 
-    private val compositeDisposable = CompositeDisposable()
+    /* symbol meaning:
+    >> : bind direction left to right, future changes are propagated
+    << : bind direction right to left, future changes are propagated
+    >| : bind direction left to right, future changes are not propagated
+    |< : bind direction right to left, future changes are not propagated
+     */
 
-    private val clickRelay = PublishRelay.create<User>()
-    private val firstNameRelay = BehaviorRelay.create<String>()
-    private val lastNameRelay = BehaviorRelay.create<String>()
+    // view |< >> view model
+    var firstName: String by distinctObservable(
+            initialValue = "",
+            onChange = firstNameRelay::onNext)
 
-    init {
-        compositeDisposable.add(userEditModel.results(clickRelay, firstNameRelay, lastNameRelay)
-                .subscribe(this::handleResult, this::handleError))
-    }
+    // view |< >> view model
+    var lastName: String by distinctObservable(
+            initialValue = "",
+            onChange = lastNameRelay::onNext)
 
-    fun setSaveClicks(obs: Observable<Unit>) {
-        compositeDisposable.add(obs
-                .map { User(firstNameRelay.value, lastNameRelay.value) }
-                .subscribe(clickRelay))
-    }
+    // view << view model
+    @get:Bindable
+    var firstNameError: String by bindable("")
+        private set
```

Comment:
The property can now only be changed internally, this is not possible with `ObservableField` etc.